tiledb/sm/misc/resource_pool.h

```diff
@@ -139,6 +144,94 @@ class ResourcePool {
   friend class ResourceHandle<T, tiledb::sm::ResourcePool>;
 };
 
+template <class T>
+class BlockingResourcePool {
+ public:
+  using resource_handle = ResourceHandle<T, tiledb::sm::BlockingResourcePool>;
+
+  /** Constructor. */
+  BlockingResourcePool(unsigned int n)
+      : resources_(n)
+      , unused_(n)
+      , unused_idx_(n - 1)
+      , num_blocked_threads_(0) {
+    for (unsigned int i = 0; i < n; i++)
+      unused_[i] = i;
+  }
+
+  /** Take a resource from the pool. */
+  resource_handle take() {
+    if (resources_exhausted()) {
+      // block until available again
+      std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
+      num_blocked_threads_++;
+      exhaustion_cv_.wait(e_lck, [this]() { return !resources_exhausted(); });
+      num_blocked_threads_--;
+    }
+    return add_resource();
+  }
+
+ private:
+  /** Release a resource from the pool. */
+  void release(unsigned int n) {
+    remove_resource(n);
+    if (blocked_threads())
+      exhaustion_cv_.notify_one();
+  }
+
+  /** Access a resource from the internal vector. */
+  T& at(unsigned int n) {
+    return resources_.at(n);
+  }
+
+  /** Check if resource pool is full */
+  inline bool resources_exhausted() {
+    return (unused_idx_ == -1);
+  }
+
+  /** Check if there is any blocked thread */
+  inline bool blocked_threads() {
+    std::unique_lock<std::mutex> e_lck(exhaustion_mtx_);
+    return (num_blocked_threads_ > 0);
+  }
+
+  /** Remove a resource from the pool */
+  inline void remove_resource(unsigned int n) {
+    std::lock_guard m_lck(m_);
+    unused_[++unused_idx_] = n;
```

Comment:
FWIW, Might just be me, but the semantics/naming you've used feel 'flipped', if 'unused_' is the pool, then it seems to me 'remove_resource()' is stating the opposite of what's being done, return_resource()/release_resource() would seem more descriptive of what's happening...?
And with 'release()', that -does- match what I'd expect, but then it calls 'remove_resource()', which just based on names wouldn't be doing what I'd expect from the name...
And the opposite operation add_resource(), I'd expect to be 'adding' to the pool/unused_, the action of current add_resource() seems more like an obtain_resource()/acquire_resource()...

Reply:
I eliminated those functions that were creating confusion (and cumbersome locking)